### ISSUE TITLE
CodableClosure should use Decodable -> Encodable

### ISSUE THREAD

**Issue by papierschiff** [IBM-Swift/Kitura#1235]
## Context and Description
At the moment the CodableClosure forces the users that the request-object and the response-object is Codable. So if the response is only Encodable, it is not possible to use the Codable-Route.

### Environment Details
MacOS 10.13.3 (17D102)

Apple Swift version 4.0.3 (swiftlang-900.0.74.1 clang-900.0.39.2)
XCode Version 9.2 (9C40b)
Kitora Version 2.2.0

### Steps to Reproduce
1) 1. create a codable-route, for example
```
router.post("/users", handler: { (request: MyRequest, respondWith: (MyResponse?, RequestError?) -> Void) in
    respondWith(nil, nil)
})
```
2) MyRequest implements Decodable, MyResponse implements Encodable
3) Doesn't compile
4) Adding Codable to MyRequest and MyResponse will work

**Comment by ianpartridge** [comment]
I think this is a reasonable request but wouldn't it be a breaking API change? In other words, wouldn't it have to wait for Kitura 3?

**Comment by djones6** [comment]
I'm not entirely convinced this would be breaking: the Codable functions of the Router class are public but the class itself is not `open`.  If it were open, you'd have a problem of Router subclasses no longer overriding the same function signatures, but that's currently not possible (outside of the Kitura package itself).

From a regular Kitura user's perspective, since Encodable and Decodable are a subset of Codable, this is effectively a relaxing of requirements on those types, so AFAIK it wouldn't break source compatibility.

But, I might be missing something.  A coded up prototype would be easier to reason about.

We'd also want to mirror the change in KituraKit.

**Comment by seabaylea** [comment]
Yes, this is actually a loosing of requirement so shouldn't be a breaking change - which means we can deliver under a semver minor.

**Comment by ianpartridge** [comment]
@papierschiff we have some draft PRs up to resolve this - can you take a look and confirm this is what you had in mind? Thanks!

**Comment by djones6** [comment]
FYI, this has been merged into master, and will be tagged in the next minor release of Kitura.

**Comment by djones6** [comment]
Unfortunately, this makes the types declared on Codable routes incompatible with the TypeDecoder (which relies on the ability to introspect a type via Decodable), so we have had to back it out for now.  I'm reopening this issue.
